File: Examples/ObjC/Sources/Analytics/CheckoutViewController.m
#import "CheckoutViewController.h"
#import "AppDelegate.h"

@interface CheckoutViewController ()

- (double)priceForProduct:(NSString *)productId;

@end

@implementation CheckoutViewController

- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];
  [AppDelegate.braze logCustomEvent:@"open_checkout_controller"
                         properties:@{
                           @"checkout_id" : self.checkoutId,
                           @"product_ids" : self.productIds,
                         }];
}

- (void)userDidPurchaseProduct:(NSString *)productId {
  double price = [self priceForProduct:productId];
  [AppDelegate.braze logPurchase:productId
                        currency:@"USD"
                           price:price
                      properties:@{@"checkout_id" : self.checkoutId}];
}

- (double)priceForProduct:(NSString *)productId {
  NSArray<NSNumber *> *prices = @[ @0.5, @8.0, @14.99, @0, @999.999 ];
  return prices[arc4random_uniform((uint32_t)prices.count)].doubleValue;
}

@end

File: Examples/Swift/Sources/Analytics/CheckoutViewController.swift
import UIKit

class CheckoutViewController: UIViewController {

  /// The internal checkout identifier
  var checkoutId: String = ""

  /// The list of identifiers for the products to checkout
  var productIds: [String] = []

  override func viewDidLoad() {
    super.viewDidLoad()

    self.navigationItem.leftBarButtonItem = UIBarButtonItem(
      barButtonSystemItem: .done, target: self, action: #selector(doneButtonSelected)
    )
  }

  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    AppDelegate.braze?.logCustomEvent(
      name: "open_checkout_controller",
      properties: [
        "checkout_id": checkoutId,
        "product_ids": productIds,
      ]
    )
  }

  @objc func doneButtonSelected() {
    self.dismiss(animated: true)
  }

  func userDidPurchase(productId: String) {
    let price = self.price(productId: productId)
    AppDelegate.braze?.logPurchase(
      productId: productId,
      currency: "USD",
      price: price,
      properties: ["checkout_id": checkoutId]
    )
  }

  private func price(productId: String) -> Double {
    [0.5, 8.0, 14.99, 0, 999.999].randomElement()!
  }

}
